### PR TITLE
added snapdragon rc pwm driver

### PR DESCRIPTION
**Pull request by tumbili**
This branch can be used to fly snapdragon with Pixfalcon / Pixracer or Pixhawk as companion for rc input and pwm output.
You can flash this branch onto the Snapdragon AND the Pixfalcon.
You will have to configure e.g. the Pixfalcon such that it suits the vehicle you are flying. This is because the mixing is happening on the Pixfalcon and not on the Snapdragon.
This step allowed much easier integration of the two systems.
Also make sure to calibrate the rc on both the Pixfalcon and the Snapdragon. The vehicle can be armed as usual with the safety switch and the rc control.

Comment: The actuator commands are sent form Snapdragon to Pixfalcon using UART at 115200 bit/s. It appears that this link is too slow to run at 250 Hz, therefore there is a usleep(5000) in uart_esc.cpp at line 474. We need so see if we can run the UART at a higher rate.
https://github.com/PX4/Firmware/blob/rc_pwm/src/drivers/uart_esc/uart_esc.cpp#L474

The UART on the Pixfalcon which is used to communicate with Snapdragon is specified here
https://github.com/PX4/Firmware/blob/rc_pwm/ROMFS/px4fmu_common/init.d/rcS#L817
Make sure that this replicates the port you are using. This is tested with the Pixfalcon!

You will have to create a serial cable with connects the ~~ESC connector J12~~ J13 to the serial output of the Pixfalcon (telem output).

This branch was tested on a quad in the lab, see video below.

https://www.youtube.com/watch?v=spBVfxyKi3Q&feature=youtu.be

Summary:
1) Flash this branch to Pixfalcon using make px4fmu_v2_default upload
2) Flash this branch onto the Snapdragon using the commands in the devguide
3) Push the startup scripts onto the Snapdragon (they have been modified)

4) Do the complete setup of the Pixfalcon via QGC as if you wanted it to fly the quad you are using.
This is important because otherwise Pixfalcon will not arm and so no pwm is output.
5) Do the rc calibration on the Snapdragon after you have connected the system together.

6) Start mainapp on the snapdragon
7) Switch safety off on the Pixfalcon and try to arm the system via rc.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Impressive PR documentation!

**Comment by LorenzMeier** [comment]
Impressive PR docs!

File: cmake/configs/nuttx_px4fmu-v2_default.cmake [hunk 1]
```diff
@@ -46,6 +46,7 @@ set(config_module_list
 	drivers/pwm_input
 	drivers/camera_trigger
 	drivers/bst
+	drivers/snapdragon_rc_pwm
```
Comment (LorenzMeier) [inline cmake/configs/nuttx_px4fmu-v2_default.cmake]: Is there a specific benefit of using a separate new file over just running the MAVLink app as we do for offboard control anyway?

**Comment by LorenzMeier** [comment]
The overall implementation makes sense and its great to see a proper MAVLink interface! We still have some architectural work to do, since our goal is use a single link implementation, not just replace the previous encoding with MAVLink encoding. I think there are three things that we likely will want to improve:

- Switch to 921600 baud. There should be nothing holding us back from using a slower baud rate.
- We should use the same API as for offboard control on Pixhawk, because this is essentially the very same use case. This means using the normal MAVLink app on the Pixhawk side and publishing the actuator controls there. The code for this is already present.
- Enable offboard mode by default in a new config for Pixhawk, but allow to override it with the offboard switch if configured. This will allow us to use Pixhawk as safety net if the main flight solution fails.

All of this is already present in the current architecture, so it's merely a question of integration and testing.

**Comment by LorenzMeier** [comment]
If you add a new config (instead of editing the main startup script on Pixhawk) we could merge this state as-is and replace the standalone app with the default MAVLink app once we're done. As this already works it would be good to get it in.

**Comment by tumbili** [comment]
@julianoes thanks for keeping this going!

**Comment by julianoes** [comment]
@tumbili no worries, I'm currently working on integrating ekf2. Once that's done, I'll consolidate these changes.

File: ROMFS/px4fmu_common/init.d/rcS [hunk 2]
```diff
@@ -807,6 +809,17 @@ else
 	mavlink start -r 800000 -d /dev/ttyACM0 -m config -x
 fi
 
+# XXX Do not run any mavlink instances since we need the serial port for communication with snapdragon
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/rcS]: This should be done according to this guide instead:
http://dev.px4.io/airframes-adding-a-new-frame.html

**Comment by LorenzMeier** [comment]
@julianoes @tumbili Once the autostart is done according to this guide instead of hardcoded this would be good to go:
http://dev.px4.io/airframes-adding-a-new-frame.html

The driver app should be build for FMUv2 and v4 (Pixfalcon and Pixracer).

**Comment by julianoes** [comment]
@LorenzMeier I'm on this and I'll get to the autostart, thanks for the link.

**Comment by julianoes** [comment]
@LorenzMeier:
> We should use the same API as for offboard control on Pixhawk, because this is essentially the very same use case. This means using the normal MAVLink app on the Pixhawk side and publishing the actuator controls there. The code for this is already present.

Which mavlink message do you suggest?

**Comment by LorenzMeier** [comment]
ACTUATOR_CONTROL_TARGET. But not right now. Right now we should just make this a little cleaner / easier to use in the startup config and get it in.

**Comment by julianoes** [comment]
Ok, I just realized that the uart_esc driver is actually made for the Qualcomm 4in1 ESCs but has now been bastardized for mavlink communication. I think I should either split the code in two modules, or just delete all Qualcomm specifc stuff because I have no way to test it.

**Comment by LorenzMeier** [comment]
Delete all Qualcomm specific stuff. The Qualcomm ESC doesn't really exist in the wild, and there is no point for us right now in supporting it.

**Comment by julianoes** [comment]
Ok, will do, and if we ever need it again we'll go back in time :).

File: ROMFS/px4fmu_common/init.d/rcS [hunk 3]
```diff
@@ -804,9 +807,42 @@ then
 	# Try to get an USB console
 	nshterm /dev/ttyACM0 &
 else
-	mavlink start -r 800000 -d /dev/ttyACM0 -m config -x
+	# Don't start mavlink in passthrough mode because we currently can't seem to stop it.
+	if [ $PASSTHROUGH == no ]
+	then
+		mavlink start -r 800000 -d /dev/ttyACM0 -m config -x
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/rcS]: This is pretty evil as it will make it impossible to reconfigure the board. Is there an issue with a MAVLink app running on USB in the background?

File: posix-configs/eagle/flight/px4.config [hunk 4]
```diff
@@ -3,17 +3,17 @@ qshell start
...
 sleep 1
+param set MAV_TYPE 2
 df_mpu9250_wrapper start
 df_bmp280_wrapper start
 df_hmc5883_wrapper start
 sensors start
 commander start
-attitude_estimator_q start
-position_estimator_inav start
+ekf2 start
```
Comment (LorenzMeier) [inline posix-configs/eagle/flight/px4.config]: Is the use of EKF2 here intentional?

Comment (julianoes) [inline posix-configs/eagle/flight/px4.config]: Since you already merged the ekf2 branch, it should already be in there.

File: src/drivers/uart_esc/uart_esc.cpp [hunk 5]
```diff
@@ -171,313 +122,341 @@ int mixer_control_callback(uintptr_t handle,
...
+	for (unsigned i = _outputs.noutputs; (i < actuator_outputs_s::NUM_ACTUATOR_OUTPUTS)
+	     && (i < actuator_controls_s::NUM_ACTUATOR_CONTROLS); ++i) {
+		controls_message.controls[i] = NAN;
 	}
 
-	_fd = fd;
-	return 0;
+	controls_message.time_usec = _controls.timestamp;
+
+	const uint8_t msgid = MAVLINK_MSG_ID_ACTUATOR_CONTROL_TARGET;
```
Comment (LorenzMeier) [inline src/drivers/uart_esc/uart_esc.cpp]: This is way too complex. I will replace this with the proper API calls.

File: ROMFS/px4fmu_common/init.d/rcS [hunk 6]
```diff
@@ -804,9 +840,14 @@ then
 	# Try to get an USB console
 	nshterm /dev/ttyACM0 &
 else
-	mavlink start -r 800000 -d /dev/ttyACM0 -m config -x
+	# Don't start mavlink in passthrough mode because we currently can't seem to stop it.
```
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/rcS]: Why is this needed?

File: posix-configs/eagle/flight/px4.config [hunk 7]
```diff
@@ -3,17 +3,17 @@ qshell start
...
 land_detector start multicopter
 mc_pos_control start
 mc_att_control start
-uart_esc start -d /dev/tty-2
+uart_esc start -d /dev/tty-3
```
Comment (LorenzMeier) [inline posix-configs/eagle/flight/px4.config]: Is tty-3 the "ESC port"?

File: ROMFS/px4fmu_common/init.d/rcS (hunk 6 above)
Comment (julianoes) [inline ROMFS/px4fmu_common/init.d/rcS]: mavlink stop-all fails for the USB one. Therefore don't start it at all... It's a hack.

File: posix-configs/eagle/flight/px4.config (hunk 7 above)
Comment (julianoes) [inline posix-configs/eagle/flight/px4.config]: Yes it is. Should be J13

File: ROMFS/px4fmu_common/init.d/rcS (hunk 6 above)
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/rcS]: It will make upgrading impossible (other than just on plugging it in). This is not something we can merge.

Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/rcS]: I fixed it in 6d42595. Please rebase and remove this hack.

File: posix-configs/eagle/flight/px4.config (hunk 7 above)
Comment (julianoes) [inline posix-configs/eagle/flight/px4.config]: Not true actually, but I'll fix it.

**Comment by julianoes** [comment]
I've pushed the changes that were needed to get this to fly. There are various issues remaining but stabilized flight was possible. 

I think the current changes don't break anything else, so we can merge this to master.

Since I was using an old Pixracer as IO board for which params don't work, I needed the following hack locally:

```
diff --git a/ROMFS/px4fmu_common/init.d/rcS b/ROMFS/px4fmu_common/init.d/rcS
index 97b98fe..12ba08e 100644
--- a/ROMFS/px4fmu_common/init.d/rcS
+++ b/ROMFS/px4fmu_common/init.d/rcS
@@ -87,6 +87,12 @@ then
 	fi
 
 	#
+	# XXX: Hack to have passthrough mode on old Pixracer boards
+	#
+	param set SYS_AUTOSTART 20000
+	param save
+
+	#
 	# Start system state indicator
 	#
 	if rgbled start
```

File: ROMFS/px4fmu_common/init.d/rcS (hunk 6 above)
Comment (julianoes) [inline ROMFS/px4fmu_common/init.d/rcS]: Done 22c42bb61fde4ddf55cdc40967e7a6eecc4e7d5a

File: posix-configs/eagle/flight/px4.config (hunk 7 above)
Comment (julianoes) [inline posix-configs/eagle/flight/px4.config]: Done 170f4e0a5daab5564da93e2d57b712c4bd666cac

**Comment by LorenzMeier** [comment]
Merged, thanks @tumbili and @julianoes
